### ISSUE TITLE
aggregate too slow?

### ISSUE THREAD

**Issue by Matrix-Zhang** [mongodb-labs/mongo-rust-driver-prototype#232]
i use aggregate to query documents where should have 110 results, i use iter to map the results to some type, elapsed 5 secs, how to optimize？

**Comment by saghm** [comment]
Hi! Can you provide an example to reproduce this (self-contained code to run, contents of database to run against, etc.)?

**Comment by Matrix-Zhang** [comment]
```
extern crate bson;
extern crate mongodb;
extern crate serde;
#[macro_use]
extern crate serde_derive;
extern crate serde_json;

use std::time::SystemTime;

use bson::{Bson, to_bson};
use mongodb::{Client, ThreadedClient};
use mongodb::db::ThreadedDatabase;
use serde_json::Value;

#[derive(Debug, Deserialize, Serialize)]
struct Model {
    id: String,
    logs: Vec<i32>,
}

fn main() {
    let client = Client::connect("localhost", 27017).expect("Failed to initialize client.");
    let collection = client.db("demo").collection("models");

    let mut documents = vec![];

    for i in 0..100 {
        let model = Model {
            id: i.to_string(),
            logs: vec![1, 2, 3, 4, 5],
        };

        documents.push(to_bson(&model).unwrap().as_document().unwrap().to_owned());
    }

    collection.insert_many(documents, None).unwrap();

    let project =
        doc!{
        "$project" => {
            "id" => 1,
            "last_log" => {
                "$slice" => ["$logs", (-1)]
            }
        }
    };

    let mut values: Vec<Value> = vec![];

    let time = SystemTime::now();

    let cursor = collection.aggregate(vec![project], None).unwrap();

    for result in cursor {
        if let Ok(document) = result {
            values.push(Bson::from(document).into());
        }
    }

    println!(
        "values's count: {}, elapsed: {}s",
        values.len(),
        time.elapsed().unwrap().as_secs()
    );

    collection.drop().unwrap();
}
```
the output is :
values's count: 100, elapsed: 4s

**Comment by saghm** [comment]
Hi again, and sorry about the delay! It turns out there's a slight bug that was introduced in bea146b92a81d40288bcb7e20a3a6d6a34145b4b; when we switched to setting the default batch size of 0 to allow the server to choose, it also caused aggregation to switch to fetching only one document at a time (due to aggregation setting the batch size with a different options where the server does _not_ interpret 0 as meaning "server default batch size"). Luckily, this is really easy to fix! I'll submit a PR right after sending this comment.

**Comment by saghm** [comment]
Fixed in 0.3.4. Thanks for reporting the issue!
